### ISSUE TITLE
Observing High variance accuracy in DL4J

### ISSUE THREAD

**Issue by rahul-raj** [deeplearning4j/deeplearning4j#5343]
I converted a logistic regression problem from keras to dl4j. I have 10000 data sets out of which 8000 sets used for training. 

Here is the data set:
https://github.com/rahul-raj/Problem-Solving/blob/master/Churn_Modelling.csv 

Here is what I did in Keras:

1) Split the entire data set into vectors ->  `X_Train`, `X_Test`, `Y_Train`, `Y_Test` where `X` correspond to input labels and `Y` correspond to output.  First three labels : `RowNumber, CustomerId & Surname` are ignored. 
2) Transform & scale them, then feed to neural network.
3) Find confusion matrix at the end.

Here is the python code:
https://github.com/rahul-raj/Problem-Solving/blob/master/ANN.py

Ignore lines 71 to 79 -> code for exporting keras model.

My DL4J code looks like this:
https://github.com/rahul-raj/Problem-Solving/blob/master/DeepLearning4j.java

I just tried to make it exactly the same implementation as Keras. Only the difference is that I used datasets instead of `INDArray` (data pre-processing) , because otherwise I can't make use of `TransformProcessRecordReader` in the code.

But I guess that doesn't make an issue here. When I use imported model from keras, I was able to see confusion matrix same as in keras. 
https://github.com/rahul-raj/Problem-Solving/blob/master/Test.java

So, now I'm left with only one option, that is to import model from keras and use it on dl4j everytime. I would like to perform everything on dl4j itself. So, I'm trying to figure out why there's high variance in the accuracy compared to the same code in Keras?  Correct me if you spot discrepancies in the code. Thank you!

@AlexDBlack

**Comment by rahul-raj** [comment]
Here's a sample execution seen from console:
```
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 9970 is 0.7496373833550347
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 9980 is 0.7416436089409723
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 9990 is 0.7460279676649305
args = [

========================Evaluation Metrics========================
 # of classes:    2
 Accuracy:        0.2770
 Precision:       0.5380
 Recall:          0.5237
 F1 Score:        0.3336
Precision, recall & F1: macro-averaged (equally weighted avg. of 2 classes)


=========================Confusion Matrix=========================
    0    1
-----------
  192 1418 | 0 = 0
   28  362 | 1 = 1
```

**Comment by AlexDBlack** [comment]
OK, finally got to looking at this, apologies for the delay.

Comparing your implementations:
DL4J: 	https://github.com/rahul-raj/Problem-Solving/blob/master/DeepLearning4j.java
Keras: 	https://github.com/rahul-raj/Problem-Solving/blob/master/ANN.py

Some differences and issues I've seen:
1. Keras config doesn't have dropout, DL4J config does
2. DL4J config is set to an inappropriate value: dropout(0.1) means 10% probability of retaining the activations (90% set to 0), which is way too high - https://deeplearning4j.org/doc/org/deeplearning4j/nn/conf/layers/Layer.Builder.html#dropOut-double-
3. Your use of Evaluation class is incorrect - just pass in the raw probabilities, don't do the thresholding before passing to Evaluation class


Also I can't speak to your KFoldIterator - does that use the same batch size as your Keras implementation?

Finally, you are not setting a learning rate. In this case, the default of 1e-3 is used (same as Keras), but still not good IMO to rely on the defaults (better to be explicit).

Anyway, maybe fix those issues and let me know if the training performance is more in line with the Keras implementation.

**Comment by rahul-raj** [comment]
Thank you for Alex for the analysis and valuable observations. Here are my comments on the issues mentioned:

Keras config doesn't have dropout, DL4J config does ->

**Adding the dropout code on keras gives a small rise to the accuracy, and can be ignored. So, I removed that from Keras code. The difference in accuracy in both the cases are less.** 

DL4J config is set to an inappropriate value: dropout(0.1) means 10% probability of retaining the activations (90% set to 0), which is way too high https://deeplearning4j.org/doc/org/deeplearning4j/nn/conf/layers/Layer.Builder.html#dropOut-double-  ->

 **I just took it in the reverse order. Now I edited this to 0.9.**

Your use of Evaluation class is incorrect - just pass in the raw probabilities, don't do the thresholding before passing to Evaluation class ->

   **I think you're referring to this line of code:
   `output = output.cond(new AbsValueGreaterThan(0.50));`
   I commented this line now**

Also I can't speak to your KFoldIterator - does that use the same batch size as your Keras implementation? ->

   **Can you guide me how to set batch size for `KFoldIterator`? I don't see a parameterized constructor for this. I have defined `batchsize` 10 for `RecordReaderDataSetIterator` in previous line of code. 
   Currently I'm seeing 8000/10 -> 800 as `batchSize` for my `KFoldIterator` and trying to figure out how to set this `batchSize` to 10.**

Finally, you are not setting a learning rate. In this case, the default of 1e-3 is used (same as Keras), but still not good IMO to rely on the defaults (better to be explicit). ->

**Acknowledging that default rate might not be good to use. We can try and experiment various learning rates to find out one that gives more accuracy. I would try passing values like this:
`.updater(new Adam(0.0002D))` 
But then, I'm seeing high variant accuracy for the same learning rate.  So, may be we can rectify high variance and then get back to this point. I will also try using regularization.**

**Comment by treo** [comment]
The KFoldIterator is rather simple, it takes only a dataset and splits it into k pieces. Given the problem you're describing in your issue, I'd suggest that you simply skip using it all together

From what I can see here, the biggest difference is probably the kfold iterator use now, because you get 9 updates (default split with kfold iterator: k = 10, as one is held out for evaluation) with a batchsize of 800 instead of 800 updates with a batchsize of 10.

**Comment by rahul-raj** [comment]
I removed the `KFoldIterator` and my training code will look like this now:
`for(int i=0;i<100;i++){
            multiLayerNetwork.fit(trainSet);
        }
`
Result is pretty much the same, still high variance :(

**Comment by rahul-raj** [comment]
@treo 
Thanks for pointing out that.
Done below changes as you mentioned:
```
DataSetIterator iterator = new RecordReaderDataSetIterator(transformProcessRecordReader,batchSize,labelIndex,numClasses);
DataSetIteratorSplitter splitter = new DataSetIteratorSplitter(iterator,10000,0.8); //10000 total no.of examples
```

```
model.setListeners(new ScoreIterationListener(100));
model.fit(splitter.getTrainIterator(),100);
```

And my execution now look like this:
```
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 99200 is 9.209391784667968
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 99300 is 6.966828918457031
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 99400 is 5.937229919433594
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 99500 is 6.906944274902344
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 99600 is 4.96789665222168
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 99700 is 8.198497009277343
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 99800 is 6.97906494140625
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 99900 is 9.218247222900391
```

But now trying to figure out how would I pass this `test iterator` to `evaluation` class so that I could print `evaluation metrics` and `confusion matrix`.

I'm getting exception while I was trying to do something like this:
```
  Evaluation evaluation = new Evaluation(1);
  INDArray output = model.output(splitter.getTestIterator()); // Exception in thread "main" org.nd4j.linalg.exception.ND4JIllegalStateException: Can't concatenate 0 arrays
```

**Comment by treo** [comment]
The full stacktrace for your exception would be more helpful here. But anyway, since you now have an iterator that you can pass around, you should probably use `model.evaluate(splitter.getTestIterator())`, it will provide you with an Evaluation instance from which you'd just output the stats.

**Comment by rahul-raj** [comment]
@treo ,

Here's the full stacktrace:
```
Exception in thread "main" org.nd4j.linalg.exception.ND4JIllegalStateException: Can't concatenate 0 arrays
	at org.nd4j.linalg.cpu.nativecpu.CpuNDArrayFactory.concat(CpuNDArrayFactory.java:660)
	at org.nd4j.linalg.factory.Nd4j.concat(Nd4j.java:5925)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.output(MultiLayerNetwork.java:2147)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.output(MultiLayerNetwork.java:2157)
	at DeepLearning4j.main(DeepLearning4j.java:82)
```

I will update the code as mentioned and post the results here. Thank you!

**Comment by rahul-raj** [comment]
@treo 
Added below one:
```
Evaluation evaluation = model.evaluate(splitter.getTestIterator());
System.out.println("args = " + evaluation.stats() + "");
```
Now I'm getting different exception, here is the full stacktrace:

```
Exception in thread "main" java.lang.UnsupportedOperationException
	at org.deeplearning4j.datasets.iterator.DataSetIteratorSplitter$2.getLabels(DataSetIteratorSplitter.java:188)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.evaluate(MultiLayerNetwork.java:3181)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.evaluate(MultiLayerNetwork.java:3102)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.evaluate(MultiLayerNetwork.java:2958)
	at DeepLearning4j.main(DeepLearning4j.java:81)
```

**Comment by treo** [comment]
That is one of the reasons why the Splitter isn't supposed to be used outside of testing. But anyway:

You can easily pass the labels from the original iterator: 
`model.evaluate(splitter.getTestIterator(), iterator.getLabels())`

**Comment by rahul-raj** [comment]
`model.evaluate(splitter.getTestIterator(), iterator.getLabels())`
I added the same. Still getting the above mentioned error :(

**Comment by treo** [comment]
Well, that means that you don't have any named labels originally. 

But you can still use that method, you just have to provide labels yourself, e.g.
`model.evaluate(splitter.getTestIterator(), Arrays.asList("1"))`

**Comment by rahul-raj** [comment]
Can't we specify input label index here rather than adding a list manually?

**Comment by treo** [comment]
The point here is that those are label names, not actual labels that you specify. So if your original dataset iterator had those, you wouldn't have seen that exception in the first place. 

What you are doing here, is exactly the same as what you did previously when you used Evaluation(1). If you want to use that instead, you can continue doing so: `model.doEvaluation(splitter.getTestIterator(), evaluation)` (with evaluation being defined as you did it earlier) should result in pretty much the same thing.

**Comment by rahul-raj** [comment]
@treo ,
But then we used datasetsplitter and passed them directly to model instead of having datasets. 
I checked if there's a away to set labels there, but couldnt find any. 
I'm a bit confused here how to get the same accuracy in my dl4j code as in my Keras code. 
Trying to make the same implementation in dl4j as in Keras.

**Comment by treo** [comment]
I could just drop the proper code on you and be done with it, but then you will still be confused. 

What you have to understand is that dl4j has disentangled the data pipeline from training the model. That means that you can't just pass all of your data off to model.fit. Usually we expect you to divide your data before even loading it, but to skip this step, you are using the DataSetIteratorSplitter. 

When you go on to run evaluation, you might as well pass all your data in at once, but you've got an iterator, so instead you can pass in an iterator. But as it doesn't have label names, you've got to either provide them yourself, or you have to let it generate them (which in turn is 0,1,2, etc... ). 

When you use `model.evaluate(iter)`, it will try to get labels from the iterator. As DataSetIteratorSplitter is meant for testing, it doesn't support the `.getLabels` call. So we tried in to pass in the labels from your main data iterator, but it didn't have any as well. As you only have two labels (0, 1), I suggested that you simply enumerate them manually. But, apparently you didn't want to do that. So instead I've shown you how you can let the Evaluation class generate those labels themselves, and then pass that one on to your model, so the evaluation can be done properly. 

So, please step back and take some time to properly understand what is happening at each line of your code. Both in your Keras implementation and in your DL4J implementation. Do not treat it as a Black Box and do not give up. Even if you decide that you just want to continue on using Keras because you don't like how DL4J does things, you will have a better understanding if how the steps work in isolation.

**Comment by rahul-raj** [comment]
@treo,
Thanks for your details explanation. 
Actually, I have tried this one when you suggested passing labels as list:
`        Evaluation evaluation = model.evaluate(splitter.getTestIterator(),Arrays.asList("0","1"));
        System.out.println("args = " + evaluation.stats() + "");
`
And I got this output:
```
========================Evaluation Metrics========================
 # of classes:    2
 Accuracy:        0.0000
 Precision:       0.0000
 Recall:          0.0000
 F1 Score:        NaN
Precision, recall & F1: reported for positive class (class 1 - "1") only

Warning: 2 classes were never predicted by the model and were excluded from average precision
Classes excluded from average precision: [0, 1]
Warning: 2 classes were never predicted by the model and were excluded from average recall
Classes excluded from average recall: [0, 1]

=========================Confusion Matrix=========================
 0 1
-----
 0 0 | 0 = 0
 0 0 | 1 = 1

Confusion matrix format: Actual (rowClass) predicted as (columnClass) N times
==================================================================
```

**Comment by treo** [comment]
This looks very much like it didn't get any test examples what so ever, as the confusion matrix is empty. Can you update your dl4j code to be the exact same code as you where using here?

Along with your iteration count going to 99900 (actually 100000, but it starts at 0, and outputs only every 100 iterations), this looks like there is either some misuse of the iterator, or the Splitter has some problems.

**Comment by rahul-raj** [comment]
Latest code updated as in my local:
https://github.com/rahul-raj/Problem-Solving/blob/master/DeepLearning4j.java

**Comment by treo** [comment]
Yes, that is why it is printing every 100 iterations, but that isn't the main point. The main Point is that apparently the train iterator gets all 10000 examples instead of getting just 8000 examples and leaving 2000 examples for testing. So at this point I'll have to take a closer look at what causes this weird behavior.

**Comment by rahul-raj** [comment]
@treo
In the meantime I will also check this on my local. Thank you!

**Comment by treo** [comment]
It looks like the problem there is very simple: DataSetIteratorSplitter requires the number of **batches** the underlying iterator has, not the total number of examples, ever though the constructor parameter is called totalExamples. So if you set it to 1000 ( = 10000 Examples / 10 Examples per Batch), it will work correctly.

**Comment by rahul-raj** [comment]
@treo 
Thanks a lot, that was truly amazing :)
I just ran the code and here are the results and observations:
```
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 79000 is 0.5136202812194824
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 79100 is 0.5961973190307617
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 79200 is 1.0605329513549804
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 79300 is 0.6056034564971924
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 79400 is 0.7506316661834717
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 79500 is 0.4980428218841553
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 79600 is 0.5240177154541016
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 79700 is 0.6336871147155761
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 79800 is 0.49535531997680665
[main] INFO org.deeplearning4j.optimize.listeners.ScoreIterationListener - Score at iteration 79900 is 0.45655322074890137
args = 

========================Evaluation Metrics========================
 # of classes:    2
 Accuracy:        0.8050
 Precision:       0.8050
 Recall:          0.5000
 F1 Score:        0.0000
Precision, recall & F1: reported for positive class (class 1 - "1") only

Warning: 1 class was never predicted by the model and was excluded from average precision
Classes excluded from average precision: [1]

=========================Confusion Matrix=========================
    0    1
-----------
 1610    0 | 0 = 0
  390    0 | 1 = 1

Confusion matrix format: Actual (rowClass) predicted as (columnClass) N times
==================================================================
```

So, now it iterates through exactly 80000 times as expected. I get an accuracy of 80.50 % consistently for various learning rates (0.01, 0.001, 0.0001 etc) now. If we change the learning rate that we pass to adam, does it have effect on the accuracy? What I think is "yes", but in that case I should see some changes in the accuracy (either increase/decrease) here right? I'm getting 80.50% for all the learning rates I passed to adam. Sorry, If I'm bumping you too much, but this just got into my mind.

**Comment by treo** [comment]
At this point you are just looking at a plain old tuning problem. One thing that you definitely should do is add a seed to your configuration, to make the initialization random, but fixed. If you don't you will see differences between two runs simply because you ran it an additional time.

Your situation here is now that you are seeing an overfitting case: The network learns that everything is class 0.

**Comment by treo** [comment]
Do you have the same overfitting with keras? If not, do as I said in an earlier comment: Step back, and try to find out what parameters you have in keras (since some are implicit, it may be a good idea to export the model and take a look at the json file's contents). Then try to set the exactly same parameters in DL4J.

**Comment by rahul-raj** [comment]
Here is what I could see on my json model from Keras:
```
{
   "class_name":"Sequential",
   "config":[
      {
         "class_name":"Dense",
         "config":{
            "name":"dense_31",
            "trainable":true,
            "batch_input_shape":[
               null,
               11
            ],
            "dtype":"float32",
            "units":6,
            "activation":"relu",
            "use_bias":true,
            "kernel_initializer":{
               "class_name":"RandomUniform",
               "config":{
                  "minval":-0.05,
                  "maxval":0.05,
                  "seed":null
               }
            },
            "bias_initializer":{
               "class_name":"Zeros",
               "config":{

               }
            },
            "kernel_regularizer":null,
            "bias_regularizer":null,
            "activity_regularizer":null,
            "kernel_constraint":null,
            "bias_constraint":null
         }
      },
      {
         "class_name":"Dense",
         "config":{
            "name":"dense_32",
            "trainable":true,
            "units":6,
            "activation":"relu",
            "use_bias":true,
            "kernel_initializer":{
               "class_name":"RandomUniform",
               "config":{
                  "minval":-0.05,
                  "maxval":0.05,
                  "seed":null
               }
            },
            "bias_initializer":{
               "class_name":"Zeros",
               "config":{

               }
            },
            "kernel_regularizer":null,
            "bias_regularizer":null,
            "activity_regularizer":null,
            "kernel_constraint":null,
            "bias_constraint":null
         }
      },
      {
         "class_name":"Dense",
         "config":{
            "name":"dense_33",
            "trainable":true,
            "units":1,
            "activation":"sigmoid",
            "use_bias":true,
            "kernel_initializer":{
               "class_name":"RandomUniform",
               "config":{
                  "minval":-0.05,
                  "maxval":0.05,
                  "seed":null
               }
            },
            "bias_initializer":{
               "class_name":"Zeros",
               "config":{

               }
            },
            "kernel_regularizer":null,
            "bias_regularizer":null,
            "activity_regularizer":null,
            "kernel_constraint":null,
            "bias_constraint":null
         }
      }
   ],
   "keras_version":"2.1.5",
   "backend":"tensorflow"
}
```

There's a min-max range for the weights, but other than that nothing specific. I had this model imported for a previous issue tracker here. The only difference in dl4j code is that we are not setting min-max range for weights, but indeed specifying random uniformity here: `weightInit(WeightInit.UNIFORM)`

**Comment by treo** [comment]
One thing that I've noticed is that you are using a very low learning rate and that as the data isn't balanced, it does have a tendency to favor the majority class. By increasing the batch size (note you'll also have to decrease the Spliterator Batch count in that case, or simply set it to 10000 / batchSize), that effect can be somewhat countered, but to get really good results you'd need some more changes to architecture and hyperparameters (see also https://deeplearning4j.org/visualization). 

Can you share what kind of accuracy, F1 and if possible also MCC you're getting with your keras model?

**Comment by rahul-raj** [comment]
In Keras, I'm observing below set of accuracy on 10-fold training set:
0.835
0.83375
0.835
0.82375
0.8475
0.85
0.835
0.83
0.81875
0.8425

I get similar accuracy everytime when I run Keras code.

And here's the score obtained after passing test set through the model:
**Error:** 0.4064310743808746
**Score**: 0.8385

And here's the confusion matrix from Keras:
  
        0              1
	 
0    1560	       35

1     288	      117

I tried changing the splitter batch size and reduced the total batch count accordingly. But no luck. In fact, a lot of permutations and combinations tried, however outcome was same. 
 
https://github.com/rahul-raj/Deeplearning4J/blob/master/src/main/java/DeepLearning4j.java

I will run DeepLearning4j UI on my local and will post the snap soon. Hopefully we can figure out from there.

**Comment by rahul-raj** [comment]
@treo,
Here are the links to UI snapshots:
https://i.imgur.com/Rnha0S9.jpg
https://www.photobox.co.uk/my/photo/full?photo_id=500824070289

Sorry about the links instead of direct attachments.
Currently on restricted network and snaps taken using mob.

**Comment by rahul-raj** [comment]
@treo 
Here we go, UI snapshots from a training instance:
![image](https://user-images.githubusercontent.com/517415/41122454-3a105ac0-6ab9-11e8-9314-1bc65638f925.png)
![image](https://user-images.githubusercontent.com/517415/41122479-4d7c3192-6ab9-11e8-9622-a40223cf9fbc.png)

However observing some exception messages on console:
https://gist.github.com/rahul-raj/c9a25a0488ed2be83750fb9fbc3408dc

**Comment by rahul-raj** [comment]
@treo @AlexDBlack 
Done plenty of testing after yesterday's discussion. Here are the things which I have tried out:

**1) Softmax with 2 inputs, LossMCXENT loss func** -> Thanks to @treo for pointing out this. Softmax was certainly the last option my mind, but data imbalance was the actual culprit here. 

**2) Changing learning rates alone** -> Observed better results in higher LR in the order of 1e-2 for my use case. As soon as the output activation is changed and loss function is applied, changes in learning rate had big influence on results. I was observing an accuracy increase of 5-6%.

**3) Different updaters with different learning rates** -> adam, rmsprop, nesterovs (sgd with momentum) -> After a quick search, found that **sgd with momentum** would be best option and slight increase of accuracy is noticed while testing.

**4) L2 regularization**  -> Optimal rate found to be 0.0005) for my use case.

**5) Added another dense layer with 8 neurons**. -> Not much of gain if layers are increased. But I just added one more deep layer.

**6) Increase epochs and dropouts together**. -> Increasing epochs from 100 to 2000 (with batchsize 10, totalBatches=1000) resulted in 1.6 million iterations. Further increase of Dropouts doesn't really helped out even though it's a lot of iterations. Just referred a forum post from Jeremy (fast.ai) where he suggests larger epochs with more dropouts (upto 0.7). 

**7) Increase/decrease epochs and learning rate together**. ->
No noticeable changes here.

At the end, after all such permutations and combinations of trials, received an avg.accuracy of 85.5% throughout testing and seems consistent though (observed range: 84.8-86.3)

F1 score observed to be between 48-56%
![image](https://user-images.githubusercontent.com/517415/41201229-7c26eca2-6cd1-11e8-9697-bddb05ac76dc.png)
![image](https://user-images.githubusercontent.com/517415/41201235-876dba0a-6cd1-11e8-8ef3-8f2b866457d4.png)

I'm closing this since the high variance has been stopped and observing a good accuracy overall. 
Correct me if any of my above mentioned procedures are incorrect or if doesn't have any significance. 
Accordingly, I will continue to try optimizing things, although it might be the case where I need more train data to get more accuracy. I hope you don't mind posting questions/comments here later if any. Thanks everyone!
